### PR TITLE
Compile-time detection of non-serializable frames

### PR DESCRIPTION
**Pull request by pleroy**
The litmus test is `apsides_test.cpp`, which uses a non-serializable frame and construct complex objects with it, and survived to tell the story.

### CODE REVIEW

File: base/traits.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace base {
+namespace internal_traits {
+
+template<typename, typename = void, typename = void>
+struct has_templated_read_message : std::false_type {};
+
+template<typename T>
+struct has_templated_read_message<
+    T, std::void_t<decltype(&T::template ReadFromMessage<>)>>
+    : std::true_type {};
```
Comment (eggrobin) [inline base/traits.hpp]: `has_templatized_read_from_message`; but really we need the template parameters to all be defaulted, i.e., used for SFINAE; so perhaps `has_enabled_read_from_message`?

File: base/traits.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+template<typename, typename = void, typename = void>
+struct has_vanilla_read_message : std::false_type {};
```
Comment (eggrobin) [inline base/traits.hpp]: `has_unconditional_read_from_message`.

File: base/traits.hpp (hunk 1 above)
Comment (pleroy) [reply base/traits.hpp]: Calling it `has_sfinae_read_from_message` because that's what it is.

File: base/traits.hpp (hunk 2 above)
Comment (pleroy) [reply base/traits.hpp]: Done.

**Pull request merged**